packages/playwright-test/src/worker/testInfo.ts

```diff
@@ -109,11 +109,19 @@ export class TestInfoImpl implements TestInfo {
 
   _deadlineForMatcher(timeout: number): { deadline: number, timeoutMessage: string } {
     const startTime = monotonicTime();
-    const matcherDeadline = timeout ? startTime + timeout : MaxTime;
+    const matcherDeadline = timeout ? startTime + timeout : 0;
     const testDeadline = this._timeoutManager.currentSlotDeadline() - 100;
     const matcherMessage = `Timeout ${timeout}ms exceeded while waiting on the predicate`;
     const testMessage = `Test timeout of ${this.timeout}ms exceeded`;
-    return { deadline: Math.min(testDeadline, matcherDeadline), timeoutMessage: testDeadline < matcherDeadline ? testMessage : matcherMessage };
+    const deadline = this._calculateEarliestDeadline(matcherDeadline, testDeadline);
+    return { deadline, timeoutMessage: deadline === matcherDeadline ? matcherMessage : testMessage };
```

Comment:
Now there is just a small chance of the wrong error message - when `matcherDeadline === testDeadline`.